tests/MongoDB.Bson.Tests/Jira/CSharp1559Tests.cs

```diff
@@ -0,0 +1,201 @@
+﻿/* Copyright 2020-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Reflection;
+using FluentAssertions;
+using MongoDB.Bson.Serialization;
+using MongoDB.Bson.Serialization.Attributes;
+using MongoDB.Bson.TestHelpers;
+using Xunit;
+
+namespace MongoDB.Bson.Tests.Jira
+{
+    public class CSharp1559Tests
+    {
+        [Theory]
+        [InlineData(typeof(DerivedWithoutSetter_BaseWithoutSetter))]
+        [InlineData(typeof(DerivedWithoutSetter_BaseWithPrivateSetterAndWithProtectedConstructor))]
+        [InlineData(typeof(DerivedWithoutSetterAndWithBsonElement_BaseWithoutSetterAndWithProtectedConstructor))]
+        [InlineData(typeof(DerivedWithoutSetterAndWithBsonElementAndWithPrivateConstructor_BaseWithoutSetter))]
+        [InlineData(typeof(DerivedWithoutSetterAndWithoutBsonElement_AbstractBaseWithoutSetterAndWithProtectedConstructor))]
+        [InlineData(typeof(DerivedWithoutSetterAndWithoutBsonElement_AbstractBaseWithoutSetter))]
+        [InlineData(typeof(DerivedWithPrivateSetterAndWithBsonElement_BaseWithoutSetter))]
+        public void Serialization_should_work_as_expected(Type testCaseType)
+        {
+            var testCase = Activator.CreateInstance(testCaseType, 1, 2);
+            var json = testCase.ToJson();
+            var result = BsonSerializer.Deserialize(json, testCaseType);
+            var x = GetPropertyValue(result, "X");
+            var y = GetPropertyValue(result, "Y");
+            x.Should().Be(1);
+            y.Should().Be(2);
+        }
+
+        [Fact]
+        public void Serialization_with_mismatching_type_between_constructor_argument_and_base_property()
```

Comment:
This test fails because the _int_ type of _x_ argument in the constructor of the derived class doesn't match with the _int?_ type of _X_ in the base class